internal/pkg/encryption/keys/kms.go

```diff
@@ -0,0 +1,165 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package keys
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/tls"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/siderolabs/go-blockdevice/blockdevice/encryption"
+	"github.com/siderolabs/go-blockdevice/blockdevice/encryption/luks"
+	"github.com/siderolabs/go-blockdevice/blockdevice/encryption/token"
+	"github.com/siderolabs/go-retry/retry"
+	"github.com/siderolabs/kms-client/api/kms"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/status"
+
+	"github.com/siderolabs/talos/internal/pkg/endpoint"
+	"github.com/siderolabs/talos/internal/pkg/smbios"
+)
+
+// KMSToken is the userdata stored in the partition token metadata.
+type KMSToken struct {
+	SealedData []byte `json:"sealedData"`
+}
+
+// KMSKeyHandler seals token using KMS service.
+type KMSKeyHandler struct {
+	key
+	kmsEndpoint string
+}
+
+// NewKMSKeyHandler creates new KMSKeyHandler.
+func NewKMSKeyHandler(key key, kmsEndpoint string) (*KMSKeyHandler, error) {
```

Comment:
the public API with private type looks a bit weird, not sure if this should be moved into yet another subpackage or `Key` should be public instead?